src/libutils/oglapphelpers/metalapp.mm

```diff
@@ -0,0 +1,391 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright Contributors to the OpenColorIO Project.
+
+
+#include <iostream>
+#include <sstream>
+#include <utility>
+#include <fstream>
+
+#import <Metal/Metal.h>
+#import <AppKit/AppKit.h>
+
+#ifdef __APPLE__
+
+#include <OpenGL/gl.h>
+#include <GLUT/glut.h>
+
+#elif _WIN32
+
+#include <GL/glew.h>
+#include <GL/glut.h>
+
+#else
+
+#include <GL/glew.h>
+#include <GL/gl.h>
+#include <GL/glut.h>
+
+#endif
+
+#include <OpenColorIO/OpenColorIO.h>
+
+#include "metalapp.h"
+#include "msl.h"
+#include "glsl.h"
+#include "mtltexture.h"
+
+namespace OCIO_NAMESPACE
+{
+
+struct GraphicsContext
+{
+    GraphicsContext() = delete;
+    GraphicsContext(NSOpenGLContext* glContext, id<MTLDevice> metalDevice)
+    {
+        this->glContext = glContext;
+        this->metalDevice = metalDevice;
+    }
+    
+    NSOpenGLContext* glContext;
+    id<MTLDevice>    metalDevice;
+};
+
+MetalApp::MetalApp(const char * winTitle, int winWidth, int winHeight)
+    : ScreenApp(winTitle, winWidth, winHeight)
+{
+    initContext();
+}
+
+void MetalApp::initContext()
+{
+    NSOpenGLContext* currentContext = [NSOpenGLContext currentContext];
+    
+    m_context = std::unique_ptr<GraphicsContext>(new GraphicsContext(currentContext, MTLCreateSystemDefaultDevice()));
+}
+
+MetalApp::~MetalApp()
+{
+    m_metalBuilder.reset();
+}
+
+void MetalApp::initImage(int imgWidth, int imgHeight, Components comp, const float * image)
+{
+    std::vector<float> imageData;
+    if(comp == Components::COMPONENTS_RGB)
+    {
+        imageData = RGB_to_RGBA(image, 3 * imgWidth * imgHeight);
+        image = imageData.data();
+    }
+    
+    setImageDimensions(imgWidth, imgHeight, comp);
+    m_image = std::make_shared<MtlTexture>(m_context->metalDevice, imgWidth, imgHeight, image);
```

Comment:
[Question] The `image` pointer is dangling when leaving the method because it's the internal pointer to a local variable i.e. `imageData`. Does `MtlTexture` copy the data?

Reply:
internally `replaceRegion` will copy the image data immediately to the texture. There is no async operation happening here.